feat: add -check_connection option for publish, add last_flush variable

piotr-karas opened the pull request:
Hi, 
I have implemented two minor changes - check what you feel about them :)

1. add last_flush variable to connection object: we have use case, where we publish some messages (e.g. every interval) and connection is lost and in this moment our process is restarted - this way we don't know if all messages were flushed to NATS server or not (well, probably not). In that case we would want to re-publish those messages that weren't send, but we don't know which one were published and which not. Using `last_flush` would allow to know what messages have really been send (comparing simple timestamp or setting trace). This way we can store messages to be re-published on disk.
2. add -check_connection to publish method - setting it to false would allow to call this method even if connection was not established yet - so in code it would look nicer, without need to store messages somewhere and publish them after connection. They would just be added to buffor and send when server is available.

Kazmirchuk commented:
just for my better understanding:
1. "connection is lost and in this moment our process is restarted" - so, NATS' usual capability to retry a connection or connect to another server in the cluster can't be used in this case? probably, because Internet is lost for a long time, and you really have to stop the whole process?
2. Really good idea! what if this becomes a new option for "configure"? so that you don't need to repeat it with every "publish" . Also, what do you think if "subscribe" behaves in the same way too with this option?

piotr-karas commented:
1. Yes, so for example after few minutes of lost connection there is process restart and all messages in buffor are lost. If we had that variable we could store those messages by ourselves on disk and after process restarts - send them again. This way we could make our mechanism bulletproof. (I don't know if it is the best solution, but it could be usefull sometimes I guess - if you don't like it let me know)
2. Good idea :) I didn't want to make too much changes, because I didn't know if you would accept that. But if you think this approach is ok I will add this to configuration method as well as to `subscribe` and probably `request` (so `configure` would save default value, and user could override it by passing explicit argument as it is now in `publish` method).
I am on vacation now so I will do it next week :)

Kazmirchuk commented:
1. is my understanding correct, that you need guaranteed delivery? Can you publish to JetStream for that? if you use the sync version of "jetstream publish", you don't need to have any separate storage for messages, compare timestamps etc. I'm afraid, relying on Tcl's "flush" is still not 100% bulletproof (but ofc better than nothing).
Are there other NATS clients that provide similar API? I can't remember any
I'm always up for new ideas, but first and foremost I'd like to help you build a resilient system.
2. I'd say just make it a `configure` option (default false), no need for extra args (unless you need to mix 2 approaches in the same program? unlikely). And check it in `publish` and `subscribe`. `request` doesn't need any special care, because it works on top of `publish`

piotr-karas commented:
1. To be honest I also thought about using JetStreams before, but was asked to make some changes so that buffer (outBuffer) management would be possible (so I have done as little changes as possible that would enable checking what is in that buffer - maybe it really wasn't the best idea). But I have convinced others to use JetStream instead, so those changes were removed.
2. I have added `-check_connection` option to `configure` method without additional arguments to publish/subscribe methods (default is `true` for backward compatibility).

(3) In last commit I have also fixed little bug in `jet_stream.tcl` (key `type` may not exists in response).

Kazmirchuk commented:
ok, I've reviewed the commit, and have only one question: why did you remove the check for
```
if {$status != $nats::status_reconnecting} 
```
I think, this check is important to avoid subscribing twice, when restoring a connection to NATS - look at `method RestoreSubs`

piotr-karas commented:
Yes, you are right - I have fixed it (and change this check to be equal to `$nats::status_connected`, because if status was `$nats::status_closed` or `$nats::status_connecting` it will be restored too if I think correctly)

Kazmirchuk commented:
well spotted, thanks
as for managing the buffer, technically you can still access it as ` ${conn}::outBuffer`. TclOO doesn't really protect "private" variables. I do it in extra_opts.test
